### PR TITLE
Issue 35: Modularization Prototype for Envision Rendering Code

### PR DESCRIPTION
**Pull request by aurorahcx**
Camera component + Vehicles component

### CODE REVIEW

**Review by Gamenot — APPROVED** [review]
Please remove the commented out lines before merge.

**Comment by liamchzh** [comment]
A good start to refactor and split things into different components.

So for the remaining parts of the `Simulation`, do they have a clear boundary between different components?

File: envision/web/src/components/simulation.js [hunk 1]
```diff
@@ -849,6 +547,17 @@ export default ({ simulationId, client, showScores, egoView, canvasRef }) => {
...
+        scene={scene}
+        roadNetworkBbox={roadNetworkBbox}
+        egoView={egoView}
+      />
+      <Vehicles
+        scene={scene}
+        worldState={worldState}
+        vehicleRootUrl={`${client.endpoint.origin}/assets/models/`}
+        egoView={egoView}
+      />
```
Comment (aurorahcx) [inline envision/web/src/components/simulation.js:560]: Camera + Vehicles modularization

**Comment by aurorahcx** [comment]
> A good start to refactor and split things into different components.
> 
> So for the remaining parts of the `Simulation`, do they have a clear boundary between different components?

Yes, I have defined the rest of components in https://github.com/huawei-noah/SMARTS/issues/35

**Review by liamchzh — APPROVED** [review]

File: envision/web/src/components/components.js [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+import { ActorTypes } from "../enums.js";
+import { intersection, difference } from "../math.js";
+import {
+  vehicleMeshFilename,
+  vehicleMeshColor,
+  buildLabel,
+} from "../render_helpers.js";
+
+export function Camera({ scene, roadNetworkBbox, egoView }) {
```
Comment (JVillella) [inline envision/web/src/components/components.js:47]: Can we move this into a sep file (that was one of the motivators of this refactoring)

File: envision/web/src/components/components.js [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+  const vehicleMeshTemplatesRef = useRef({});
+  const vehicleMeshesRef = useRef({});
+  const agentLabelGeometryRef = useRef({});
+
+  let vehicleMeshTemplates = vehicleMeshTemplatesRef.current;
+  let vehicleMeshes = vehicleMeshesRef.current;
+  let agentLabelGeometry = agentLabelGeometryRef.current;
+
+  // Load mesh asynchronously
+  useEffect(() => {
```
Comment (JVillella) [inline envision/web/src/components/components.js:148]: To improve readability, let's do a slight restructure of the code in this useEffect to leverage some methods.

File: envision/web/src/components/components.js [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+          rootMesh.setBoundingInfo(new BoundingInfo(rootMeshMin, rootMeshMax));
+
+          vehicleMeshTemplates[vehicleFilename] = rootMesh;
+        }
+      );
+    }
+    // This useEffect is triggered when the vehicleMeshTemplate's keys() change
+  }, [Object.keys(vehicleMeshTemplates).sort().join("-")]);
+
+  useEffect(() => {
```
Comment (JVillella) [inline envision/web/src/components/components.js:205]: Dup. too much code in this useEffect, hard to follow along.

File: envision/web/src/components/components.js [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    if (!meshesLoaded(vehicleMeshTemplates)) {
+      return;
+    }
+
+    let nextVehicleMeshIds = new Set(Object.keys(worldState.traffic));
+    let vehicleMeshIds = new Set(Object.keys(vehicleMeshes));
+
+    let vehicleMeshIdsToRemove = difference(vehicleMeshIds, nextVehicleMeshIds);
```
Comment (JVillella) [inline envision/web/src/components/components.js:213]: Not for this PR: This diffing technique could be abstracted out into a helper or other structure for later reuse.

**Review by JVillella — APPROVED** [review]
Few small things, the direction is good. Once we have these more modular components we'll need some successive rounds of simplification for each.

File: envision/web/src/components/components.js [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+  for (const [_, mesh] of Object.entries(vehicleMeshTemplates)) {
+    if (mesh == null) {
+      return false;
+    }
+  }
+  return true;
+};
+
+// Vehicles and agent labels
+export function Vehicles({ scene, worldState, vehicleRootUrl, egoView }) {
```
Comment (JVillella) [inline envision/web/src/components/components.js:134]: Sep. file.
